### PR TITLE
support transaction isolation modification through reserved connection system settings

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR moves the transaction isolation system variables (transaction_isolation or set transaction isolation level) from vitess aware to reserved connection.
Earlier it used to only work with transactions. With this change, support will be added for read queries.

## Related Issue(s)

- Closes https://github.com/vitessio/vitess/issues/11810
- Related to https://github.com/vitessio/vitess/pull/11673 and https://github.com/vitessio/vitess/pull/11804

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [X] Tests were added or are not required
-   [X] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/safe_session.go [hunk 1]
```diff
@@ -597,10 +597,10 @@ func (session *SafeSession) SetPreQueries() []string {
 	first := true
 	for _, k := range keys {
 		if first {
-			preQuery.WriteString(fmt.Sprintf("set @@%s = %s", k, sysVars[k]))
+			preQuery.WriteString(fmt.Sprintf("set %s = %s", k, sysVars[k]))
```
Comment (GuptaManan100) [inline go/vt/vtgate/safe_session.go:600]: Why is this change needed?

**Review by GuptaManan100 — COMMENTED** [review]
Everything else looks good, but isn't this like a breaking change? Previously setting the transaction isolation would only apply to the next query, but now it applies to the session, if I understand correctly. Shouldn't this be added to the release notes?

File: go/vt/vtgate/safe_session.go (hunk 1 above)
Comment (harshit-gangal) [reply go/vt/vtgate/safe_session.go:600]: TLDR; previously it was broken. 

@@ without any scope represents `session` scope. Only for the case of `transaction_isolation` it means the next transaction. This gets applied only once. 
With the implementation of `settings pool` these settings will not be applied again and the connection will be reused,  which means the next query the connection will see, will have the session level isolation level and not the next transaction isolation level.
So, the user expectation will turn out to be wrong.

The RFC already states that the next transaction isolation statements will be converted to the session level setting.

File: go/test/endtoend/vtgate/reservedconn/sysvar_test.go [hunk 2]
```diff
@@ -427,18 +427,36 @@ func TestSysVarTxIsolation(t *testing.T) {
...
 	utils.AssertMatches(t, conn, "select @@transaction_isolation", `[[VARCHAR("REPEATABLE-READ")]]`)
+	// ensuring it goes to mysql
+	utils.AssertContains(t, conn, "select @@transaction_isolation, connection_id()", `REPEATABLE-READ`)
+	// second run, to ensuring the setting is applied on the session and not just on next query after settings.
+	utils.AssertContains(t, conn, "select @@transaction_isolation, connection_id()", `REPEATABLE-READ`)
 
 	// setting to different value.
 	utils.Exec(t, conn, "set @@transaction_isolation = 'read-committed'")
-	utils.AssertMatches(t, conn, "select @@transaction_isolation", `[[VARCHAR("READ-COMMITTED")]]`)
+	utils.AssertMatches(t, conn, "select @@transaction_isolation", `[[VARCHAR("read-committed")]]`)
```
Comment (deepthi) [inline go/test/endtoend/vtgate/reservedconn/sysvar_test.go:441]: Why are some values uppercase and some lowercase?
Line 437 - `REPEATABLE-READ`. But everything else is lowercase.

Comment (deepthi) [reply go/test/endtoend/vtgate/reservedconn/sysvar_test.go:441]: Looks like it may be an artifact of how the test input is provided. To avoid confusion, my suggestion is to make them all uniform.

**Review by deepthi — APPROVED** [review]
One small suggestion, rest LGTM

**Review by GuptaManan100 — APPROVED** [review]
I don't have any other comments than the ones already answered.

File: go/test/endtoend/vtgate/reservedconn/sysvar_test.go (hunk 2 above)
Comment (harshit-gangal) [reply go/test/endtoend/vtgate/reservedconn/sysvar_test.go:441]: I will raise a separate PR to handle it.
